### PR TITLE
Add a note about performant record_stream use.

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112526

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

File: torch/_tensor_docs.py [hunk 1]
```diff
@@ -4039,6 +4039,37 @@ def callable(a, b) -> number
     unexpectedly. Calling this method lets the allocator know which streams
     have used the tensor.
 
+.. warning::
+
+    This method is most suitable for use cases where you are a providing a
```
Comment (albanD) [inline torch/_tensor_docs.py:4044]:
```suggestion
    This method is most suitable for use cases where you are providing a
```

File: torch/_tensor_docs.py [hunk 2]
```diff
@@ -4039,6 +4039,37 @@ def callable(a, b) -> number
...
+    and manual memory management), so if you are in a situation where
+    you manage the full lifetime of your tensors, you may consider instead
+    manually managing CUDA events so that calling this method is not necessary.
+    In particular, when you call this method, on later allocations the
+    allocator will poll the recorded stream to see if all operations have
+    completed yet; you can potentially race with side stream computation and
+    non-deterministically reuse or fail to reuse memory for an allocation.
+
+    You can safely use tensors allocated on side streams without
+    :method:`~Tensor.record_stream`; the idea is to manually ensure that that
```
Comment (albanD) [inline torch/_tensor_docs.py:4058]:
```suggestion
    :method:`~Tensor.record_stream`; the idea is to manually ensure that
```

File: torch/_tensor_docs.py [hunk 3]
```diff
@@ -4039,6 +4039,37 @@ def callable(a, b) -> number
...
+    ``del`` the tensor.  This is not a mechanical change: synchronizing immediately
+    after you queue work on the side stream would defeat the purpose of having
+    a side stream at all.  Instead, the synchronization must be placed at some
+    appropriate, later point in time where you expect the side stream to have
+    finished work.  This point could be identified via profiling.  Note that
+    this approach still implies a device-to-host synchronization, so if the
+    side stream has not actually finished work, you will block until it is
+    finished.  Conversely, if you place the deallocation too late, you will use
+    more memory than is strictly necessary.  For a concrete example of how
+    this guidance can be applied in practice, see this post: `FSDP and
```
Comment (albanD) [inline torch/_tensor_docs.py:4070]: This is not the recommendation from this blogpost actually.
The recommendation is to use `Event.wait(stream)` API to ensure that the user stream syncs with the allocation stream before the del.
This has the same caveat that this needs to be done later in time but does not have the device-to-host caveat mentioned here.

**Review by albanD — COMMENTED** [review]
two spaces after a `.` is a new trend I don't know about? :D

File: torch/_tensor_docs.py (hunk 3 above)
Comment (awgu) [reply torch/_tensor_docs.py:4070]: Do you mean `stream.wait_event(event)` (or is that the same thing)?

Comment (albanD) [reply torch/_tensor_docs.py:4070]: It is the same thing from the other side yes.

Comment (ezyang) [reply torch/_tensor_docs.py:4070]: Oh, this must be the thing where, on CPU side, we can reallocate a tensor for another stream before the original stream, so long as writes to the tensor on the new stream wait for the original stream to finish first. OK, I'll update this accordingly.

File: docs/source/notes/cuda.rst [hunk 4]
```diff
@@ -264,7 +264,29 @@ used.  For example, the following code is incorrect::
...
 However, when using non-default streams, it is the user's responsibility to
-ensure proper synchronization.
+ensure proper synchronization.  The fixed version of this example is::
+
+    cuda = torch.device('cuda')
+    s = torch.cuda.Stream()  # Create a new stream.
+    A = torch.empty((100, 100), device=cuda).normal_(0.0, 1.0)
+    s.wait_stream(torch.cuda.default_stream(cuda))  # NEW!
+    with torch.cuda.stream(s):
+        B = torch.sum(A)
```
Comment (colesbury) [inline docs/source/notes/cuda.rst:274]: This is missing a `wait_stream` or `record_stream` call. The "deallocation" of A on the default stream needs to happen *after* the `torch.sum()` reads from A. Or, in more concrete terms, without additional synchronization, the caching allocator may re-use the memory for `A` once it goes out of scope *before* the `torch.sum()` call completes.

(You explain the need for this synchronization in the `record_stream` comment)

```
    cuda = torch.device('cuda')
    s = torch.cuda.Stream()  # Create a new stream.
    A = torch.empty((100, 100), device=cuda).normal_(0.0, 1.0)
    s.wait_stream(torch.cuda.default_stream(cuda))  # NEW!
    with torch.cuda.stream(s):
        B = torch.sum(A)
    torch.cuda.default_stream(cuda).wait_stream(s)
```

or

```
    cuda = torch.device('cuda')
    s = torch.cuda.Stream()  # Create a new stream.
    A = torch.empty((100, 100), device=cuda).normal_(0.0, 1.0)
    s.wait_stream(torch.cuda.default_stream(cuda))  # NEW!
    with torch.cuda.stream(s):
        B = torch.sum(A)
A.record_stream(s)  # location of this call does not really matter
```

File: docs/source/notes/cuda.rst [hunk 5]
```diff
@@ -264,7 +264,29 @@ used.  For example, the following code is incorrect::
...
+
+Note that this synchronization is necessary even when there is no
+read dependency, e.g., as seen in this example::
+
+    cuda = torch.device('cuda')
+    s = torch.cuda.Stream()  # Create a new stream.
+    A = torch.empty((100, 100), device=cuda)
+    s.wait_stream(torch.cuda.default_stream(cuda))  # STILL REQUIRED!
+    with torch.cuda.stream(s):
+        A.normal_(0.0, 1.0)
```
Comment (colesbury) [inline docs/source/notes/cuda.rst:284]: This is also missing a necessary synchronization.

File: docs/source/notes/cuda.rst (hunk 4 above)
Comment (ezyang) [reply docs/source/notes/cuda.rst:274]: oh so embarrassing, thanks for catching

**Review by albanD — APPROVED** [review]
SGTM

**Pull request closed without merging**